widget/table.go

```diff
@@ -0,0 +1,301 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+// Table widget is a grid of items that can be scrolled and a cell selected.
+// It's performance is provided by caching cell templates created with NewCell and re-using them with UpdateCell.
+// The size of the content rows/columns is returned by the DataSize callback.
+type Table struct {
+	BaseWidget
+
+	DataSize       func() (int, int)
+	NewCell        func() fyne.CanvasObject
+	UpdateCell     func(row int, col int, template fyne.CanvasObject)
+	OnCellSelected func(row int, col int)
+
+	SelectedRow, SelectedColumn int
+	cells                       *tableCells
+	moveCallback                func()
+}
+
+// NewTable returns a new performant table widget defined by the passed functions.
+// The first returns the data size in rows and columns, second parameter is a function that returns cell
+// template objects that can be cached and the third is used to apply data at specified data location to the
+// passed template CanvasObject.
+func NewTable(size func() (int, int), create func() fyne.CanvasObject, update func(int, int, fyne.CanvasObject)) *Table {
+	t := &Table{DataSize: size, NewCell: create, UpdateCell: update, SelectedRow: -1, SelectedColumn: -1}
+	t.ExtendBaseWidget(t)
+	return t
+}
+
+// CreateRenderer returns a new renderer for the table.
+//
+// Implements: fyne.Widget
+func (t *Table) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	marker1 := canvas.NewRectangle(theme.PrimaryColor())
+	marker2 := canvas.NewRectangle(theme.PrimaryColor())
+
+	template := t.NewCell()
+	cellSize := template.MinSize().Add(fyne.NewSize(theme.Padding()*2, theme.Padding()*2))
+	t.cells = newTableCells(t, cellSize)
+	scroll := NewScrollContainer(t.cells)
+
+	obj := []fyne.CanvasObject{marker1, marker2, scroll}
+	r := &tableRenderer{t: t, scroll: scroll, rowMarker: marker1, colMarker: marker2, objects: obj, cellSize: cellSize}
+	t.moveCallback = r.moveIndicators
+	scroll.onOffsetChanged = r.moveIndicators
+
+	r.Layout(t.Size())
+	return r
+}
+
+type tableRenderer struct {
+	t *Table
+
+	scroll               *ScrollContainer
+	rowMarker, colMarker *canvas.Rectangle
+	dividers             []fyne.CanvasObject
```

Comment:
Is there any advantage to splitting dividers into horizontal and vertical so the driver doesn't have to re-draw if the size is the same?

Reply:
I don't think so - a simple filled rectangle is essentially a 1x1px uniform that is drawn at whatever size inside the GL code